File: python/lib/castling.py
from dataclasses import dataclass

from .enums import Color

@dataclass
class CastlingRights:
    white_kingside: bool = True
    white_queenside: bool = True
    black_kingside: bool = True
    black_queenside: bool = True
    
    def copy(self) -> 'CastlingRights':
        return CastlingRights(
            self.white_kingside,
            self.white_queenside,
            self.black_kingside,
            self.black_queenside
        )
    
    def get_for_color(self, color: Color):
        if color == Color.WHITE:
            return (self.white_kingside, self.white_queenside)
        else:
            return (self.black_kingside, self.black_queenside)
    
    def to_string(self) -> str:
        s = ""
        if self.white_kingside: s += "K"
        if self.white_queenside: s += "Q"
        if self.black_kingside: s += "k"
        if self.black_queenside: s += "q"
        return s if s else "-"
    
    @staticmethod
    def from_string(s: str) -> 'CastlingRights':
        cr = CastlingRights(False, False, False, False)
        if s == "-":
            return cr
        cr.white_kingside = 'K' in s
        cr.white_queenside = 'Q' in s
        cr.black_kingside = 'k' in s
        cr.black_queenside = 'q' in s
        return cr

File: rust/src/lib/castling.rs
use crate::enums::Colorr;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CastlingRights {
    pub white_kingside: bool,
    pub black_kingside: bool,
    pub white_queenside: bool,
    pub black_queenside: bool,
}

impl CastlingRights {
    pub fn new() -> Self {
        Self {
            white_kingside: true,
            black_kingside: true,
            white_queenside: true,
            black_queenside: true,
        }
    }

    pub fn castling_mask(&self) -> u8 {
        let mut mask = 0;

        if self.white_kingside {
            mask |= 1;
        } // Bit 0
        if self.white_queenside {
            mask |= 2;
        } // Bit 1
        if self.black_kingside {
            mask |= 4;
        } // Bit 2
        if self.black_queenside {
            mask |= 8;
        } // Bit 3

        mask
    }

    pub fn get_for_color(&self, color: &Colorr) -> (bool, bool) {
        if color == &Colorr::WHITE {
            return (self.white_kingside, self.white_queenside);
        } else {
            return (self.black_kingside, self.black_queenside);
        }
    }

    pub fn to_string(&self) -> String {
        let mut castlings = String::new();

        if self.white_kingside {
            castlings += "K"
        }

        if self.white_queenside {
            castlings += "Q"
        }

        if self.black_kingside {
            castlings += "k"
        }

        if self.black_queenside {
            castlings += "q"
        }

        if castlings.is_empty() {
            return String::from("-");
        }

        return castlings;
    }

    pub fn from_string(string: String) -> CastlingRights {
        CastlingRights {
            white_kingside: string.find('K').is_some(),
            black_kingside: string.find('k').is_some(),
            white_queenside: string.find('Q').is_some(),
            black_queenside: string.find('q').is_some(),
        }
    }
}
